### PR TITLE
Always delete existing scheduled action tasks

### LINKED ISSUE #27650: Old  Scheduled Action Tasks Not Cleared When Creating Remove Scheduled Task Configuration Commits
### Description

After I remove the existing cron configuration, gitea does not remove the scheduled tasks from the database,

image

Configuration parsing fails and ends without clearing the old cron task.

image

https://github.com/go-gitea/gitea/blob/0d55f64e6cd3de2e1e5c0ee795605823efb14231/services/actions/notifier_helper.go#L149-L152

### Gitea Version

main

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

systemd

### Database

SQLite

### PR DESCRIPTION
**Pull request by KN4CK3R**
Fixes #27650

### CODE REVIEW

File: services/actions/notifier_helper.go [hunk 1]
```diff
@@ -400,6 +395,11 @@ func handleSchedules(
 		}
 	}
 
+	if len(detectedWorkflows) == 0 {
```
Comment (delvh) [inline services/actions/notifier_helper.go:395]: Not your fault, but I just noticed a huge memory/performance "leak" in the code above this line while trying to understand why it fixes the problem.
I think it is quite wasteful to query and discard all schedules, when a `Has` check would suffice.

**Review by delvh — APPROVED** [review]

File: services/actions/notifier_helper.go (hunk 1 above)
Comment (KN4CK3R) [reply services/actions/notifier_helper.go:395]: Could use the `CountSchedules` method instead.

**Comment by delvh** [comment]
Something I'm not sure about:
Are our cron schedules relative or absolute?
So, does `* /5 * * *` mean ``run at multiples of `5 minutes` ``, or does it mean `` run at multiples of `5 minutes from now` ``?
If it is the latter, we will now get somewhat buggy behavior as the existing schedule is cleared.

**Comment by KN4CK3R** [comment]
The behaviour should not be different than before. If the commit changed something unrelated to the scheduled action, all existing tasks are removed and re-added.

This PR just forces the deletion even if the config does not contain a scheduled action anymore.

**Comment by lunny** [comment]
I don't think this PR fixed #27650. The problem is when pushing an error parsed yaml, all old schedule tasks haven't been stop. So we need to check cron jobs when parsing workflow yaml file failed.

**Comment by delvh** [comment]
> I don't think this PR fixed https://github.com/go-gitea/gitea/issues/27650. The problem is when pushing an error parsed yaml, all old schedule tasks haven't been stop

Well, that is one a completely different problem and two not the critical problem as that can be easily fixed by users.
The biggest problem is fixed by this PR already-

**Comment by lunny** [comment]
I think you are right these are 2 bugs but fixed and #27650 reported two. I think this PR just fixed one of two so it should partially fix it in the description.

This PR just resolved if the workflows are deleted but not from valid to invalid.

But in #27650, the second screenshot displayed the yaml file parse failed I think which is the real reason that the old cron jobs haven't been deleted.

图片

**Review by lunny — APPROVED** [review]

**Comment by KN4CK3R** [comment]
Isn't that case covered too? If the file is invalid `DetectWorkflows` returns nothing but in the end my changed method is called which removes the scheduled actions.

**Comment by lunny** [comment]
in line 149, if parse failed, it will return error but will not invoke `handleSchedules` which is located line 191.

```go
	workflows, schedules, err := actions_module.DetectWorkflows(gitRepo, commit, input.Event, input.Payload)
	if err != nil {
		return fmt.Errorf("DetectWorkflows: %w", err)
	}
```

**Pull request merged**